### PR TITLE
Fix `da.broadcast_to` call when the fx cube has different shape than target data cube

### LINKED ISSUE #1349: `mask_landsea` fails when using 2D fx data on 4D cube data
**Minimal recipe**

```yaml
documentation:
  description: |
    Various example preprocessors.
  authors:
    - righi_mattia
  title: Test recipe for ESMValCore preprocessors.
  maintainer:
    - righi_mattia
  projects:
    - c3s-magic
datasets:
  - {dataset: bcc-csm1-1, project: CMIP5, exp: historical, ensemble: r1i1p1}
preprocessors:
  preprocessor_1:
    extract_levels:
      levels: bcc-csm1-1
      scheme: linear
    mask_landsea:
      mask_out: land
diagnostics:
  diagnostic_1: &diag
    description: Preprocessor test diagnostic.
    variables:
      ta:
        preprocessor: preprocessor_1
        mip: Amon
        start_year: 2000
        end_year: 2002
      orog:
        mip: fx
      sftlf:
        mip: fx
    scripts: null
```

**Trace (full)**
```
2021-10-08 12:44:56,533 UTC [25166] ERROR   Failed to run mask_landsea(air_temperature / (K)               (time: 36; air_pressure: 17; latitude: 64; longitude: 128)
    Dimension coordinates:
        time                             x                 -             -              -
        air_pressure                     -                 x             -              -
        latitude                         -                 -             x              -
        longitude                        -                 -             -              x
    Ancillary variables:
        land_area_fraction               -                 -             x              x
    Cell methods:
        mean                        time (20 mintues)
    Attributes:
        Conventions                 CF-1.4
        associated_files            baseURL: http://cmip-pcmdi.llnl.gov/CMIP5/dataLocation gridspecFile: gridspec_atmos_fx_bcc-csm1-1_historical_r0i0p0.nc...
        branch_time                 470.0
        cmor_version                2.5.6
        contact                     Dr. Tongwen Wu (twwu@cma.gov.cn)
        experiment                  historical
        experiment_id               historical
        forcing                     Nat Ant GHG SD Oz Sl Vl SS Ds BC OC
        frequency                   mon
        initialization_method       1
        institute_id                BCC
        institution                 Beijing Climate Center(BCC),China Meteorological Administration,China
        model_id                    bcc-csm1-1
        modeling_realm              atmos
        original_name               T
        parent_experiment           pre-industrial control
        parent_experiment_id        piControl
        parent_experiment_rip       r1i1p1
        physics_version             1
        product                     output
        project_id                  CMIP5
        realization                 1
        source                      bcc-csm1-1:atmosphere:  BCC_AGCM2.1 (T42L26); land: BCC_AVIM1.0;ocean:...
        table_id                    Table Amon (11 April 2011) 1cfdc7322cf2f4a32614826fab42c1ab
        title                       bcc-csm1-1 model output prepared for CMIP5 historical, {'mask_out': 'land'})
2021-10-08 12:44:56,682 UTC [25155] INFO    Maximum memory used (estimate): 0.7 GB
2021-10-08 12:44:56,684 UTC [25155] INFO    Sampled every second. It may be inaccurate if short but high spikes in memory consumption occur.
2021-10-08 12:44:56,685 UTC [25155] ERROR   Program terminated abnormally, see stack trace below for more information:
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/home/valeriu/miniconda3-fresh/envs/esmvaltool-stock/lib/python3.9/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/home/valeriu/ESMValCore/esmvalcore/_task.py", line 780, in _run_task
    output_files = task.run()
  File "/home/valeriu/ESMValCore/esmvalcore/_task.py", line 253, in run
    self.output_files = self._run(input_files)
  File "/home/valeriu/ESMValCore/esmvalcore/preprocessor/__init__.py", line 489, in _run
    product.apply(step, self.debug)
  File "/home/valeriu/ESMValCore/esmvalcore/preprocessor/__init__.py", line 356, in apply
    self.cubes = preprocess(self.cubes, step, **self.settings[step])
  File "/home/valeriu/ESMValCore/esmvalcore/preprocessor/__init__.py", line 302, in preprocess
    result.append(_run_preproc_function(function, item, settings))
  File "/home/valeriu/ESMValCore/esmvalcore/preprocessor/__init__.py", line 285, in _run_preproc_function
    return function(items, **kwargs)
  File "/home/valeriu/ESMValCore/esmvalcore/preprocessor/_mask.py", line 118, in mask_landsea
    fx_cube.data = da.broadcast_to(fx_cube.core_data(), cube.shape)
  File "/home/valeriu/miniconda3-fresh/envs/esmvaltool-stock/lib/python3.9/site-packages/iris/coords.py", line 740, in data
    self._values = data
  File "/home/valeriu/miniconda3-fresh/envs/esmvaltool-stock/lib/python3.9/site-packages/iris/coords.py", line 241, in _values
    self._values_dm.data = values
  File "/home/valeriu/miniconda3-fresh/envs/esmvaltool-stock/lib/python3.9/site-packages/iris/_data_manager.py", line 259, in data
    raise ValueError(emsg.format(self.shape, data.shape))
ValueError: Require data with shape (64, 128), got (36, 17, 64, 128).
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/valeriu/ESMValCore/esmvalcore/_main.py", line 479, in run
    fire.Fire(ESMValTool())
  File "/home/valeriu/miniconda3-fresh/envs/esmvaltool-stock/lib/python3.9/site-packages/fire/core.py", line 141, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "/home/valeriu/miniconda3-fresh/envs/esmvaltool-stock/lib/python3.9/site-packages/fire/core.py", line 466, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
  File "/home/valeriu/miniconda3-fresh/envs/esmvaltool-stock/lib/python3.9/site-packages/fire/core.py", line 681, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
  File "/home/valeriu/ESMValCore/esmvalcore/_main.py", line 423, in run
    process_recipe(recipe_file=recipe, config_user=cfg)
  File "/home/valeriu/ESMValCore/esmvalcore/_main.py", line 125, in process_recipe
    recipe.run()
  File "/home/valeriu/ESMValCore/esmvalcore/_recipe.py", line 1557, in run
    self.tasks.run(max_parallel_tasks=self._cfg['max_parallel_tasks'])
  File "/home/valeriu/ESMValCore/esmvalcore/_task.py", line 709, in run
    self._run_parallel(max_parallel_tasks)
  File "/home/valeriu/ESMValCore/esmvalcore/_task.py", line 752, in _run_parallel
    _copy_results(task, running[task])
  File "/home/valeriu/ESMValCore/esmvalcore/_task.py", line 775, in _copy_results
    task.output_files, task.products = future.get()
  File "/home/valeriu/miniconda3-fresh/envs/esmvaltool-stock/lib/python3.9/multiprocessing/pool.py", line 771, in get
    raise self._value
  File "/home/valeriu/miniconda3-fresh/envs/esmvaltool-stock/lib/python3.9/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/home/valeriu/ESMValCore/esmvalcore/_task.py", line 780, in _run_task
    output_files = task.run()
  File "/home/valeriu/ESMValCore/esmvalcore/_task.py", line 253, in run
    self.output_files = self._run(input_files)
  File "/home/valeriu/ESMValCore/esmvalcore/preprocessor/__init__.py", line 489, in _run
    product.apply(step, self.debug)
  File "/home/valeriu/ESMValCore/esmvalcore/preprocessor/__init__.py", line 356, in apply
    self.cubes = preprocess(self.cubes, step, **self.settings[step])
  File "/home/valeriu/ESMValCore/esmvalcore/preprocessor/__init__.py", line 302, in preprocess
    result.append(_run_preproc_function(function, item, settings))
  File "/home/valeriu/ESMValCore/esmvalcore/preprocessor/__init__.py", line 285, in _run_preproc_function
    return function(items, **kwargs)
  File "/home/valeriu/ESMValCore/esmvalcore/preprocessor/_mask.py", line 118, in mask_landsea
    fx_cube.data = da.broadcast_to(fx_cube.core_data(), cube.shape)
  File "/home/valeriu/miniconda3-fresh/envs/esmvaltool-stock/lib/python3.9/site-packages/iris/coords.py", line 740, in data
    self._values = data
  File "/home/valeriu/miniconda3-fresh/envs/esmvaltool-stock/lib/python3.9/site-packages/iris/coords.py", line 241, in _values
    self._values_dm.data = values
  File "/home/valeriu/miniconda3-fresh/envs/esmvaltool-stock/lib/python3.9/site-packages/iris/_data_manager.py", line 259, in data
    raise ValueError(emsg.format(self.shape, data.shape))
ValueError: Require data with shape (64, 128), got (36, 17, 64, 128).
2021-10-08 12:44:56,690 UTC [25155] INFO    
If you have a question or need help, please start a new discussion on https://github.com/ESMValGroup/ESMValTool/discussions
If you suspect this is a bug, please open an issue on https://github.com/ESMValGroup/ESMValTool/issues
To make it easier to find out what the problem is, please consider attaching the files run/recipe_*.yml and run/main_log_debug.txt from the output directory.
```

**Issue in a nutshell**

fx data has shape: (64, 128) and data cube has shape: (36, 17, 64, 128); in `_mask.py` l.117 (not 118 as in the trace, I added a line at pos 117 inspect the two cubes)
```python
fx_cube.data = da.broadcast_to(fx_cube.core_data(), cube.shape)
```
fails to broadcast fx data onto cube data.

### PR DESCRIPTION
**Pull request by valeriupredoi**
## Description

We spotted a bug in the landsea masking related to the broadcasting of the fx data array onto the cube to be masked data array.

Closes #1349 

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Comment by valeriupredoi** [comment]
@zklaus sorry, was working on the test, now included too, it's a cold-fix now :grin:

**Comment by valeriupredoi** [comment]
be my guest @sloosvel - cheers! Make sure you pull the latest, I just added a bunch o'changes to the integration test mask :+1: Let's agree we want this in 2.4 so let's hustle up on it, and get it ready before the day of release :beer:

**Comment by sloosvel** [comment]
```
def _preserve_fx_vars(cube, result):
    vertical_dim = set(cube.coord_dims(cube.coord(axis='z', dim_coords=True)))
    if cube.cell_measures():
        for measure in cube.cell_measures():
            measure_dims = set(cube.cell_measure_dims(measure))
            if vertical_dim.intersection(measure_dims):
                logger.warning(
                    'Discarding use of z-axis dependent cell measure %s '
                    'in variable %s, as z-axis has been interpolated',
                    measure.var_name, result.var_name)
            else:
                add_cell_measure(result, measure, measure.measure)
    if cube.ancillary_variables():
        for ancillary_var in cube.ancillary_variables():
            ancillary_dims = cube.ancillary_variable_dims(ancillary_var)   --> this should be set(cube.ancillary_variable_dims(ancillary_var) )
            if vertical_dim.intersection(ancillary_dims):
                logger.warning(
                    'Discarding use of z-axis dependent ancillary variable %s '
                    'in variable %s, as z-axis has been interpolated',
                    ancillary_var.var_name, result.var_name)
            else:
                add_ancillary_variable(result, ancillary_var)

```

And I only added tests for `climate_statistics`, they should also be added for `decadal_statistics`, `annual_statistics`, etc.

File: tests/integration/preprocessor/_mask/test_mask.py [hunk 1]
```diff
@@ -48,6 +53,8 @@ def setUp(self):
                                           bounds=[[0, 1], [1, 2], [2, 3]],
                                           units='hours')
         self.coords_spec = [(self.lats, 0), (self.lons, 1)]
```
Comment (zklaus) [inline tests/integration/preprocessor/_mask/test_mask.py:55]: Perhaps rename this to `fx_coords_spec` or `horizontal_coords_spec` or similar?

**Comment by valeriupredoi** [comment]
OK @zklaus cheers for taking a look/reviewing - I agree we can merge and then Saskia can do the other changes in a subsequent PR (bit too much for this PR methinks too). Cheers both for looking into it! :beer:

**Pull request merged**